sys/riscv/trap.c

```diff
@@ -0,0 +1,301 @@
+      break;
+
+    case SCAUSE_ILLEGAL_INSTRUCTION: {
+      uint32_t badinstr = *(uint32_t *)epc;
+      klog("Illegal instruction %08x in kernel mode!", badinstr);
+      break;
+    }
+
+    case SCAUSE_BREAKPOINT:
+      klog("No debbuger in kernel!");
+      break;
+  }
+  klog("HINT: Type 'info line *%p' into gdb to find faulty code line", epc);
+
+  panic("KERNEL PANIC!!!");
+}
+
+static void page_fault_handler(ctx_t *ctx) {
+  thread_t *td = thread_self();
+
+  unsigned code = ctx_code(ctx);
+  void *epc = (void *)_REG(ctx, PC);
+  vaddr_t vaddr = _REG(ctx, TVAL);
+
+  klog("%s at %p, caused by reference to %lx!", exceptions[code], epc, vaddr);
+
+  pmap_t *pmap = pmap_lookup(vaddr);
+  if (!pmap) {
+    klog("No physical map defined for %lx address!", vaddr);
+    goto fault;
+  }
+
+  vm_prot_t access;
+  if (code == SCAUSE_INST_PAGE_FAULT)
+    access = VM_PROT_EXEC;
+  else if (code == SCAUSE_LOAD_PAGE_FAULT)
+    access = VM_PROT_READ;
+  else
+    access = VM_PROT_WRITE;
+
+  int error = pmap_emulate_bits(pmap, vaddr, access);
+  if (!error)
+    return;
+
+  if (error == EACCES || error == EINVAL)
+    goto fault;
+
+  vm_map_t *vmap = vm_map_lookup(vaddr);
+  if (!vmap) {
+    klog("No virtual address space defined for %lx!", vaddr);
+    goto fault;
+  }
+
+  if (!vm_page_fault(vmap, vaddr, access))
+    return;
+
+fault:
+  if (td->td_onfault) {
+    /* Handle copyin/copyout faults. */
+    _REG(ctx, PC) = td->td_onfault;
+  } else if (user_mode_p(ctx)) {
+    /* Send a segmentation fault signal to the user program. */
+    sig_trap(ctx, SIGSEGV);
+  } else {
+    /* Panic when kernel-mode thread uses wrong pointer. */
+    kernel_oops(ctx);
+  }
+}
+
+/*
+ * RISC-V syscall ABI:
+ *  - a7: code
+ *  - a0-5: args
+ *
+ * NOTE: the following code assumes all arguments to syscalls are passed
+ * via registers.
+ */
+static_assert(SYS_MAXSYSARGS <= FUNC_MAXREGARGS - 1,
+              "Syscall args don't fit in registers!");
+
+static void syscall_handler(mcontext_t *uctx, syscall_result_t *result) {
+  /* TODO: eventually we should have a platform-independent syscall handler. */
+  register_t args[SYS_MAXSYSARGS];
+  register_t code = _REG(uctx, A7);
+
+  memcpy(args, &_REG(uctx, A0), sizeof(args));
+
+  if (code > SYS_MAXSYSCALL) {
+    args[0] = code;
+    code = 0;
+  }
+
+  sysent_t *se = &sysent[code];
+  size_t nargs = se->nargs;
+
+  assert(nargs <= SYS_MAXSYSCALL);
+
+  thread_t *td = thread_self();
+  register_t retval = 0;
+
+  assert(td->td_proc);
+
+  int error = se->call(td->td_proc, (void *)args, &retval);
+
+  result->retval = error ? -1 : retval;
+  result->error = error;
+}
+
+#if FPU
+static bool fpu_handler(mcontext_t *uctx) {
+  thread_t *td = thread_self();
+
+  if (td->td_pflags & TDP_FPUINUSE)
+    return false;
+
+  /*
+   * May be a FPE trap. Enable FPE usage
+   * for this thread and try again.
+   */
+  bzero(uctx->__fregs, sizeof(__fregset_t));
```

Comment:
We are using `bzero, bcopy, memcpy, memset` in our code without any convention.
We have around:
- 120x `memset`
- 220x `memcpy`
-  30x `bzero`
- 40x `bcopy`

could we unify everything?

According to various manuals:
```
The bcopy() function conforms to IEEE Std 1003.1-2001 (``POSIX.1'').  The
     IEEE Std 1003.1-2004 (``POSIX.1'') revision marked it as legacy; the IEEE
     Std 1003.1-2008 (``POSIX.1'') revision removed it from the specification.
```

Reply:
Since we took large pieces of code from *BSD that we don't feel like modifying we need to keep those functions. However I reckon it'd be wise choice not to use `bzero` and `bcopy` in our code.